Major updates to the threading model

333fred opened the pull request:
This is a major refactoring of everything async, making sure that all COM/UI interactions happen on the UI thread, and everything that does not need to use the UI thread doesn't. Basically, anything that either accesses a VS interface or casts a VS object to some other interface needs to occur on the UI thread. I've renamed async methods to use the ends with Async. There were a few other things I refactored, giving GetService a generic wrapper and making the output functions async.

333fred committed:
Major updates to the threading model, making sure that all COM/UI interactions happen on the UI thread.

ThadHouse commented on FRC-Extension/Buttons/DeployDebugButton.cs:
```diff
@@ -55,27 +57,30 @@ private void EnableAllButtons()
-                    Output.WriteLine("Connection to RoboRIO lost. Deploy aborted.");
+                    await Output.WriteLineAsync("Connection to RoboRIO lost. Deploy aborted.").ConfigureAwait(false);
+                    await ThreadHelperExtensions.SwitchToUiThread();
                     EnableAllButtons();
```
For things like this, shouldn't we put the SwitchToUiThread in the EnableAllButtons function, and make that awaitable? Seems cleaner and less duplicated code.

333fred replied:
Probably a good idea

333fred committed:
Moved button disable methods to async.

ThadHouse commented:
There were some missed function name changes in RoboRioConnection.cs. Also the deploy is deadlocking for some reason inside one of the SSH functions. If I revert that entire file back, the deadlock goes away. Will continue investigating.

ThadHouse committed:
Updates the RoboRioConnection to be class based

Will be better for unit testing in the future. Was attempting to fix a
deadlock. This did not fix it, so I think the issue is a bug in the SSH
library. Implemented a workaround.

333fred committed:
Merge pull request #1 from robotdotnet/PR37

Updates the RoboRioConnection to be class based
